File: src/main/java/frc/robot/utils/pingu/AlertPingu.java
package frc.robot.utils.pingu;

import com.ctre.phoenix6.hardware.*;
import edu.wpi.first.wpilibj.*;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import java.util.*;

import static edu.wpi.first.wpilibj.Alert.AlertType.kError;

/**
 * A subsystem for aggregating Alerts
 */
public class AlertPingu extends SubsystemBase {
  private final ArrayList<TalonFX> motors = new ArrayList<>();
  private final ArrayList<CANcoder> canCoders = new ArrayList<>();
  private final ArrayList<Alert> motorAlerts = new ArrayList<>();
  private final ArrayList<Alert> canCoderAlerts = new ArrayList<>();

  /**
   * The Singleton instance of this AlertPingu. Code should use the {@link #getInstance()} method to
   * get the single instance (rather than trying to construct an instance of this class.)
   */
  private static final AlertPingu INSTANCE = new AlertPingu();

  /**
   * Creates a new instance of this AlertPingu. This constructor is private since this class is a
   * Singleton. Code should use the {@link #getInstance()} method to get the singleton instance.
   */
  private AlertPingu() {}

  /**
   * Returns the Singleton instance of this AlertPingu. This static method should be used, rather
   * than the constructor, to get the single instance of this class. For example: {@code
   * AlertPingu.getInstance();}
   */
  @SuppressWarnings("WeakerAccess")
  public static AlertPingu getInstance() {
    return INSTANCE;
  }

  /**
   * Automatically periodically set the alerts to each of the motors
   */
  @Override
  public void periodic() {
    for (int i = 0; i < motors.size(); i++) {
      motorAlerts.get(i).set(!motors.get(i).isConnected());
    }
    for (int i = 0; i < canCoders.size(); i++) {
      canCoderAlerts.get(i).set(!canCoders.get(i).isConnected());
    }
  }

  /** Add a motor to the {@link AlertPingu} subsystem
   *
   * @param motor The motor to add
   * @param motorName The name of the motor being added
   */
  public void add(TalonFX motor, String motorName) {
    motors.add(motor);
    motorAlerts.add(new Alert("Disconnected " + motorName + " motor " + motor.getDeviceID(), kError));
  }

  /** Add a canCoder to the {@link AlertPingu} subsystem
   *
   * @param canCoder The canCoder to add
   * @param canCoderName The name of the canCoder being added
   */
  public void add(CANcoder canCoder) {
    canCoders.add(canCoder);
    canCoderAlerts.add(new Alert("Disconnected canCoder " + canCoder.getDeviceID(), kError));
  }
}

File: src/main/java/frc/robot/utils/pingu/AlertPingu.kt
package frc.robot.utils.pingu

import com.ctre.phoenix6.hardware.CANcoder
import com.ctre.phoenix6.hardware.TalonFX
import edu.wpi.first.wpilibj.Alert
import edu.wpi.first.wpilibj.Alert.AlertType.kError
import edu.wpi.first.wpilibj2.command.SubsystemBase

/**
 * Type alias for a pair consisting of a device and its corresponding alert.
 * The device can be any type, and the alert is an instance of WPILib's Alert class.
 */
typealias DeviceAlert = Pair<Any, Alert>

/**
 * FRC subsystem that monitors CTRE Phoenix 6 devices and raises alerts for disconnections.
 *
 * This singleton object tracks the connection status of TalonFX motors and CANcoders,
 * providing real-time monitoring through WPILib's alert system. It's designed for early
 * detection of CAN bus or device issues during matches and testing.
 *
 * @property devicePairs Pairs of monitored devices and their corresponding alerts
 */
object AlertPingu : SubsystemBase() {
    private val devicePairs = mutableListOf<DeviceAlert>()

    /**
     * Updates connection status for all monitored devices.
     * Called periodically by the CommandScheduler during robot operation.
     * Uses parallel processing for efficient handling of multiple devices.
     */
    override fun periodic() {
        devicePairs.stream().forEach { (device, alert) ->
            when (device) {
                is TalonFX -> alert.set(!device.isConnected)
                is CANcoder -> alert.set(!device.isConnected)
            }
        }
    }

    /**
     * Registers a TalonFX motor for connection monitoring.
     *
     * @param motor TalonFX motor instance to monitor
     * @param motorName Descriptive name for identifying the motor in Driver Station alerts
     */
    @JvmStatic
    fun add(
        motor: TalonFX,
        motorName: String,
    ) {
        val alert = Alert("Disconnected $motorName motor ${motor.deviceID}", kError)
        devicePairs.add(Pair(motor, alert))
    }

    /**
     * Registers a CANcoder for connection monitoring.
     *
     * @param canCoder CANcoder instance to monitor
     */
    @JvmStatic
    fun add(canCoder: CANcoder) {
        val alert = Alert("Disconnected CANcoder ${canCoder.deviceID}", kError)
        devicePairs.add(Pair(canCoder, alert))
    }
}
